src/appleseed/foundation/image/oiioimagefilewriter.cpp

```diff
@@ -0,0 +1,506 @@
+// THE SOFTWARE.
+//
+
+// Interface header.
+#include "oiioimagefilewriter.h"
+
+// appleseed.foundation headers.
+#include "foundation/core/exceptions/exceptionioerror.h"
+#include "foundation/image/icanvas.h"
+#include "foundation/math/vector.h"
+#include "foundation/platform/compiler.h"
+#include "foundation/utility/foreach.h"
+#include "foundation/utility/iostreamop.h"
+#include "foundation/utility/string.h"
+
+// Boost headers.
+#include "boost/filesystem/path.hpp"
+
+// Standard headers.
+#include <algorithm>
+#include <cstring>
+#include <memory>
+#include <stdexcept>
+#include <vector>
+
+namespace foundation
+{
+
+struct OIIOImageFileWriter::Impl
+{
+    std::vector<const ICanvas*>     m_canvas;
+    std::vector<OIIO::ImageSpec>    m_spec;
+};
+
+OIIOImageFileWriter::OIIOImageFileWriter(const char* filename) :
+    impl{ new Impl() }
+{
+    assert(filename);
+
+    m_filename = filename;
+
+    m_writer = OIIO::ImageOutput::create(m_filename);
+    if (m_writer == nullptr)
+    {
+        const std::string msg = OpenImageIO::geterror();
+        throw ExceptionIOError(msg.c_str());
+    }
+}
+
+OIIOImageFileWriter::~OIIOImageFileWriter()
+{
+    // Destroy the ImageOutput stucture.
+    OIIO::ImageOutput::destroy(m_writer);
+
+    delete impl;
+}
+
+size_t OIIOImageFileWriter::get_image_count() const
+{
+    return impl->m_canvas.size();
+}
+
+void OIIOImageFileWriter::append_image(const ICanvas* image)
+{
+    assert(image);
+
+    impl->m_canvas.push_back(image);
+    impl->m_spec.push_back(OIIO::ImageSpec{});
+
+    set_image_spec();
+}
+
+OIIO::TypeDesc convert_pixel_format(const PixelFormat format)
+{
+    switch (format)
+    {
+      case PixelFormatUInt8:
+          return OIIO::TypeDesc::UINT8;
+    
+      case PixelFormatUInt16:
+          return OIIO::TypeDesc::UINT16;
+    
+      case PixelFormatUInt32:
+          return OIIO::TypeDesc::UINT32;
+    
+      case PixelFormatHalf:
+          return OIIO::TypeDesc::HALF;
+    
+      case PixelFormatFloat:
+          return OIIO::TypeDesc::FLOAT;
+    
+      case PixelFormatDouble:
+          return OIIO::TypeDesc::DOUBLE;
+    
+      default:
+          return OIIO::TypeDesc::UNKNOWN;
+    }
+}
+
+void OIIOImageFileWriter::set_image_output_format(const PixelFormat output_pixel_format)
+{
+    assert(!impl->m_spec.empty());
+
+    impl->m_spec.back().set_format(convert_pixel_format(output_pixel_format));
+}
+
+void OIIOImageFileWriter::set_image_channels(
+    const size_t    channel_count,
+    const char**    channel_names)
+{
+    assert(!impl->m_spec.empty());
+    assert(channel_count > 0);
+    assert(channel_names);
+
+    OIIO::ImageSpec& spec = impl->m_spec.back();
+
+    spec.nchannels = static_cast<int>(channel_count);
+
+    for (size_t i = 0; i < channel_count; i++)
+    {
+        const char* name = channel_names[i];
+
+        spec.channelnames.push_back(name);
+
+        if (strcmp(name, "A"))
```

Comment:
That looks wrong. `strcmp()` returns 0 when strings are equal.

Reply:
Done